File: go/maybe/maybe.go
package maybe

import "gooptional"

type Maybe[A any] = gooptional.Optional[A]

func Nothing[A any]() Maybe[A] {
    return gooptional.None[A]()
}

func Just[A any](a A) Maybe[A] {
    return gooptional.Some[A](a)
}

func Pure[A any](a A) Maybe[A] {
    return Just(a)
}

func Map[A, B any](m Maybe[A], f func(A) B) Maybe[B] {
    return FlatMap(m, func (a A) Maybe[B] { return Just(f(a)) })
}

func FlatMap[A, B any](m Maybe[A], f func(A) Maybe[B]) Maybe[B] {
    if m.IsSome(){
        return f(m.Get())
    } else { return Nothing[B]() }
}

func OrElseGet[A any](x Maybe[A], f func() Maybe[A]) Maybe[A] {
    if x.IsSome(){
        return x
    } else { return f() }
}

File: PHP/src/Maybe/Maybe.php
<?php
declare(strict_types=1);

namespace MonadParser\Maybe;

use MonadParser\Maybe\Constructor\{Just, Nothing};

abstract class Maybe
{
    final public static function just($x) : Just
    {
        return new Just($x);
    }

    final public static function nothing() : Nothing
    {
        return new Nothing();
    }

    /**
     * Functor implemenattion
     */
    abstract public function map(callable $f) : Maybe;

    /**
     * Applicative implementation.
     */
    public function ap_get(callable $fthat) : Maybe
    {
        return $this->flatMap(function ($f) use ($fthat){ return $fthat()->map($f); });
    }

    public function ap(Maybe $that) : Maybe
    {
        return $this->ap_get(function () use ($that){ return $that; });
    }

    /**
     * Monad implementation.
     * PHP implemenattion of Haskell Maybe's >>=.
     */
    abstract public function flatMap(callable $f) : Maybe;

    /**
     * Alternative implementation
     */
    abstract public function orElseGet(callable $fthat) : Maybe;

    public function orElse(Maybe $that) : Maybe
    {
        return this->orElseGet(function () use ($that){ return $that; });
    }

    abstract public function equals(Maybe $that) : bool;

    /**
     * Fork this Maybe, with a default for Nothing.
     * @param mixed $default In case of Nothing.
     * @return mixed Whatever the Maybe's inner type is.
     */
    abstract public function fork($_);
}
